exonum/src/blockchain/config.rs

```diff
@@ -295,6 +296,135 @@ impl ValidateInput for ConsensusConfig {
     }
 }
 
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    Eq,
+    Hash,
+    Serialize,
+    Deserialize,
+    ProtobufConvert,
+    BinaryValue,
+    ObjectHash,
+)]
+#[protobuf_convert(source = "runtime::GenesisConfig")]
+pub struct GenesisConfig {
+    /// Blockchain configuration used to create the genesis block.
+    pub consensus_config: ConsensusConfig,
+
+    /// Artifacts specification of builtin services.
+    pub artifacts: Vec<ArtifactSpec>, // TODO: HashSet? Duplications? Perhaps check in builder?
+
+    /// List of the privileged services with the configuration parameters that are created directly
+    /// in the genesis block.
+    pub builtin_instances: Vec<ConfiguredInstanceSpec>,
+}
+
+/// A wrapper for `InstanceSpec` that also stores a constructor argument for the service instance.
+/// Used in `GenesisConfig`.
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    Eq,
+    Hash,
+    Serialize,
+    Deserialize,
+    ProtobufConvert,
+    BinaryValue,
+    ObjectHash,
+)]
+#[protobuf_convert(source = "runtime::ConfiguredInstanceSpec")]
+// TODO: Think about better name and/or substitution the `InstanceSpec`
+pub struct ConfiguredInstanceSpec {
+    /// Wrapped `InstanceSpec`.
+    pub instance_spec: InstanceSpec,
+    /// Constructor argument for specific `InstanceSpec`.
+    pub constructor: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub struct GenesisConfigBuilder {
+    /// Consensus config.
+    consensus_config: ConsensusConfig,
+    /// Artifacts specifications for builtin services.
+    artifacts: Vec<ArtifactSpec>,
+    /// Instances of builtin services.
+    builtin_instances: Vec<ConfiguredInstanceSpec>,
+}
+
+#[derive(Debug)]
+pub struct InstanceConfig {
+    pub artifact_spec: ArtifactSpec,
+    pub instances: Vec<ConfiguredInstanceSpec>,
+}
+
+impl InstanceConfig {
+    pub fn new(artifact_id: ArtifactId, deploy_args: Vec<u8>) -> Self {
+        let artifact_spec = ArtifactSpec {
+            artifact: artifact_id,
+            payload: deploy_args,
+        };
+        Self {
+            artifact_spec,
+            instances: vec![],
+        }
+    }
+
+    pub fn with_instance(
+        mut self,
+        instance_spec: InstanceSpec,
+        constructor: impl BinaryValue,
+    ) -> Self {
+        assert_eq!(self.artifact_spec.artifact, instance_spec.artifact);
+        self.instances.push(ConfiguredInstanceSpec {
+            instance_spec,
+            constructor: constructor.into_bytes(),
+        });
+        self
+    }
+
+    pub fn from_spec(instance_spec: InstanceSpec, constructor: impl BinaryValue) -> Self {
```

Comment:
By this method there is no way to set deploy args, is this a correct approach?

Reply:
Yes. The main intention of this method was using in tests for converting in tests. I wanted to reflect this in the documentation, but forgot. Now I see that this method is now such useful, so probably it will be better to get rid of it.